changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,613 @@
+
+Here, by prefix we mean, adding the UTF8 encoding
+of the string 'hierarchy_keyring' as a
+prefix to a hierarchy keyring cache identifier,
+and adding encoding of 'caching_cmm' as a prefix to a
+CachingCMM cache identifier.
+
+#### Option 1 (Recommended): Yes
+
+Note that after adding Key Store ID / partition ID to a
+Hierarchy Keyring's cache identifier, multiple Hierarchy Keyrings
+can use a shared cache. Multiple CachingCMMs can use a
+shared cache already as mentioned in the [background](#background).
+Adding prefixes is an additional safety measure to prevent any
+possible collisions between a
+Hierarchy Keyring and a CachingCMM.
+
+#### Option 2: No
+
+If we don't add prefixes, even though the chance of an
+overlap is minimal, we might get collisions between a
+Hierarchy Keyring and a CachingCMM cache identifier.
+Adding prefixes adds another layer of distinction between them.
+
+## How can we ensure cache identifiers are collision resistent?
+
+To answer this question, I will use two approaches.
+
+### Derivation from assumptions
+
+For the HKeyring and CachingCMM, ensuring that all cache
+identifiers are unique for all encrypts and decrypts
+is simple if we just study the statements in the
+[background](#background) and the fields we're
+appending to the identifiers as discussed in issues
+[1](#issue-1-how-to-update-the-cache-identifier-for-the-hierarchy-keyring-to-allow-a-shared-cache)
+and [4](#issue-4-should-we-add-a-prefix-to-all-cache-identifiers-in-hierarchy-keyrings-and-cachingcmms).
+
+Two HKeyrings WILL NOT have collisions if we append
+a Key Store ID parameter to the identifier
+(per issue
+[1](#issue-1-how-to-update-the-cache-identifier-for-the-hierarchy-keyring-to-allow-a-shared-cache)).
+
+Two CachingCMMs already allow shared cache.
+
+Per issue
+[4](#issue-4-should-we-add-a-prefix-to-all-cache-identifiers-in-hierarchy-keyrings-and-cachingcmms),
+a Hierarchy Keyring and a CachingCMM idenitifier will
+be distinct if we add the UTF8 encodings of prefixes
+'hierarchy_keyring' (17 bytes) and 'caching_cmm' (11 bytes)
+to each of them respectively.
+
+### Distinct sizes of the identifiers
+
+As an additional safety measure, we look closely
+at the formulae for cache identifiers in a Hierarchy Keyring
+and a CachingCMM encrypt and decrypt (note that these will
+actually be a total of 5, since the
+[cachingCMM encrypt cache identifier](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/caching-cmm.md#appendix-a-cache-entry-identifier-formulas)
+has two different formulae based on if we
+specify the algorithm suite or not).
+
+In the
+[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula)
+Encrypt, as discussed before, we will remove the len_branch_key.
+For Decrypt, we will remove the len_branch_key
+and use the branch-key-digest in place of the branch-key-id.
+We also add the keystore-id, which is a UUID,
```

Comment:
issue: keystore-id defaults to a UUID, but it can be user input... so it can be `n` length.

Reply:
Ohh yes you're right.

Because we want the cache identifier to be of uniform length, the best way to move forward would be to hash the Key Store ID.

I'll update this in the doc after the meeting.

Reply:
Maybe? 
Alternatively, we could "invent" a new field for the Key Store object;
key store digest.

The Key Store digest is either:
- UUID, if no Key Store ID is given
- SHA(Key Store ID) if Key Store ID is given

If we can find a SHA that gives 16 bytes, you end up with 16 bytes either way.